drake/systems/System.h

```diff
@@ -0,0 +1,242 @@
+#ifndef DRAKE_SYSTEM_H
+#define DRAKE_SYSTEM_H
+
+#include <Eigen/Core>
+#include <Eigen/Dense>
+#include "Core.h"
+#include "SystemSpecializations.h"
+
+
+/// A Drake::System is a dynamical system that is compatible with most of our tools for design and analysis.
+/// It must have:
+///   - a real-vector-valued input, state, and output
+///   - deterministic dynamics and outputs given the input and state
+///   - continuous dynamics [ coming soon: no more than one discrete time-step (in addition to continuous dynamics) ]
+///  The input, state, and output coordinate systems are all described by CoordinateSystem objects
+///  In addition, it MAY have
+///   - time-varying dynamics and outputs
+///   - input limits (c++ support coming soon)
+///   - algebraic constraints (c++ support coming soon)
+///   - zero-crossings (c++ support coming soon) to inform the tools of discontinuities in the dynamics
+
+
+namespace Drake {
+
+  template <typename Derived, template<typename> class StateVector, template<typename> class InputVector, template<typename> class OutputVector, bool _isTimeVarying = true, bool _isDirectFeedthrough = true >
+  class System {
+  public:
+    typedef Derived Type;
+    template <typename ScalarType> using StateVectorType = StateVector<ScalarType>;
+    template <typename ScalarType> using InputVectorType = InputVector<ScalarType>;
+    template <typename ScalarType> using OutputVectorType = OutputVector<ScalarType>;
+    const static bool isTimeVarying = _isTimeVarying;
+    const static bool isDirectFeedthrough = _isDirectFeedthrough;
+    const static unsigned int num_states = VectorTraits<StateVector<double> >::RowsAtCompileTime;
+    const static unsigned int num_inputs = VectorTraits<InputVector<double> >::RowsAtCompileTime;
+    const static unsigned int num_outputs = VectorTraits<OutputVector<double> >::RowsAtCompileTime;
```

Comment:
I think we want to use int instead of unsigned int. Certainly Eigen uses the same enum trick for e.g. RowsAtCompileTime, and they can have a negative number of rows at compile (-1 to indicate dynamic size). Note that the actual underlying type of an enum is not specified by the standard (https://stackoverflow.com/questions/159034/are-c-enums-signed-or-unsigned). The best way to get around it I think is to use constexpr instead of the enum trick in VectorTraits, but I believe that MSVC 2013 doesn't fully support constexpr yet.